### DISCUSSION TITLE
Recommended practices for building in-house/commercial SDKs and libraries

### DISCUSSION THREAD

**Discussion by donnie-jp** [MobileNativeFoundation/discussions#135] [Proposals]
Hi all,

In our team we build in-house SDKs. Many companies once they reach a certain scale will build SDKs to make app devs lives easier. There are also many companies who provide a mobile SDK for using their product (usually a backend service). There isn't that much written around SDK and I think it's sufficiently different from standard app dev that the MNF could create and host some recommendations around it.

I summarized ours in this [post](https://dev.to/donniejp/how-we-build-and-deliver-quality-mobile-sdks-at-scale-269k):
* High quality
  * Peer reviewed code
  * Automated unit tests that are run by CI on every PR/merge
* Easy to use
  * Simple, fluent APIs
  * Excellent user documentation
  * Responsive developer support
* Easily integrated
  * CocoaPods/Gradle dependency
  * Configured in zero-to-a-few lines of code
* Deployed by automation
  * Minimize effort
  * Minimize risk of human error
* Well-designed/architected
  * Limit dependencies
  * Only expose what must be exposed
  * Components loosely coupled
  * Robust and secure
* Properly versioned
  * Following semantic versioning
* Consistent between platforms
  * Same API interface
  * Same high-level architecture
* Maintainable long term
  * Stable after reaching v1.0
  * At least 6 months notice of end of life

Since that post we've also added SPM support.

Is this a decent list? Are yours different or do you have more recommendations to add?

Thanks
Donnie

**Comment by omarzl** [comment]
Hello, that list is pretty good!
Specifically for iOS we have strict rules for 3rd party SDKs integration in our app:
- We ask for a static linked binary, discusion about it: https://github.com/MobileNativeFoundation/discussions/discussions/133, so as an SDK provider I think you should support static and dynamic linking and let the client decide which one benefits him more
- We expect max 1 MB increment in download size, you should mention somewhere the size impact it will have
- We ask for library evolution support, we don’t want to wait for another company to move to a new Xcode version

**Reply by donnie-jp** [reply]
Those are nice additions - thanks!

We don't track download size but maybe we should.

Supporting pre-release versions of Xcode is something we already do in our roadmap - we should add it in the guidelines too 👍 

Another related thing we do is increasing the minimum supported version in the same quarter that new major OS versions are released. This is done so we can gradually take advantage of more modern OS features and also reduce our QA load.

**Comment by keith** [comment]
> Configured in zero-to-a-few lines of code

FWIW I would say avoiding `zero` lines is a good idea to avoid this kind of issue https://9to5mac.com/2020/07/10/app-crash-facebook-sdk/ since I think folks are very wary of this after these public issues.

Also I haven't watched in a while, but I remember thinking https://academy.realm.io/posts/altconf-conrad-kramer-writing-iOS-sdk/ was a good talk on this topic

**Reply by donnie-jp** [reply]
Yes, that's a good point about the Facebook issue.
Unless there's a very good reason for it, configuration/startup of the SDK should be in the control of the app so that they can use remote feature flags to disable the SDK if necessary.

Yeah the Conrad Kramer talk is a really good reference - it was one of the few posts I found when I started in an SDK team.

**Comment by donnie-jp** [comment]
Another iOS specific thing is method swizzling. 
It's a great/useful technique but I don't think it belongs in production code unless it's explicitly required for your SDK to be able to do its job. And if you do use swizzling you should provide an API, or build config flag, to disable it.

**Comment by keith** [comment]
Internally we have a checklist folks need to fill out in order to integrate third party SDKs, here are some of the bullets we check:

- Supports minimum OS deployment target
- Does not use private APIs
- Is not closed source precompiled Swift framework (this is to prevent us from being tied to the Xcode update cadence of a third party)
- Transitive dependency does not conflict with an existing dependency (i.e., 4.x vs 3.x)
- Does not display its own UI to our users
- APIs are annotated for nullability
- Does not rely on or networks with upstream APIs/endpoints that are not owned by us - or can be kill switched to ensure upstream failure does not cause us failure.
- SDK builds without warnings
- Open sourced, If Android, has proper maven package distribution (.jar/.aar , .pom, private/public maven repo) no matter if open source or not
- Has docs/changelog
- Public interfaces are documented with Javadocs/Headerdoc
- Does not swizzle
- Has a low dex count
- Has a small footprint on our binary size
- Issue reporting/escalation policy
- If Objective-C, has prefixed symbol names
- Crash rate < 0.02%
- Has permissive license (if open source)

We weight these a bit different and come up with some score that gives us a pass / fail

**Reply by donnie-jp** [reply]
> Internally we have a checklist folks need to fill out in order to integrate third party SDKs, here are some of the bullets we check:
>  <list snipped>
> We weight these a bit different and come up with some score that gives us a pass / fail

That's a strict checklist!
Interesting insight, thanks for sharing.
